File: test/test_extract_clip_reencode.cc
/**
 * (c) Chad Walker, Chris Kirmse
 */

#include <stdio.h>

#include <string>

#include "../utils.h"
#include "../video_reader.h"

#include "file_io_group.h"

void logProgress(int step, int total) {
    printf("progress %i/%i\n", step, total);
}

int main(int argc, char **argv) {
    if (argc < 5) {
        printf("Need filename to read, filename to write, start_time, and end_time\n");
        return 1;
    }

    std::string source_pathname = argv[1];
    std::string dest_pathname = argv[2];

    double start_time = std::stod(argv[3]);
    double end_time = std::stod(argv[4]);

    Avalanche::setDefaultLogFunc();

    printf("lavf version %s\n", Avalanche::getAvFormatVersionString().c_str());

    FileIoGroup file_io_group;

    Avalanche::VideoReader video_reader;

    if (!video_reader.init(&file_io_group, source_pathname)) {
        printf("video reader init failed\n");
        return 1;
    }

    if (!video_reader.verifyHasVideoStream()) {
        printf("video has no video stream\n");
        return 1;
    }

    Avalanche::ExtractClipResult clip_result;
    if (!video_reader.extractClipReencode(dest_pathname, start_time, end_time, clip_result, logProgress)) {
        printf("failed to extract clip reencode\n");
        return 1;
    }
    printf("total video packets %i key frames %i start_time %f duration %f\n", clip_result.count_video_packets, clip_result.count_key_frames, clip_result.video_start_time, clip_result.video_duration);

    return 0;
}

File: nodejs_wrapper/test_extract_clip_reencode.js
/**
 * (c) Chad Walker, Chris Kirmse
 */

import log from '../log.js';

import Avalanche from '../avalanche.js';
import ResourceIo from '../resource_io.js';

const main = async function () {
  if (process.argv.length !== 6) {
    log.info(
      'usage: test_extract_clip_reencode.js <source_filename> <dest_filename> <start_time> <end_time>',
    );
    return;
  }

  log.info('lavf version', Avalanche.getAvFormatVersionString());

  const source_uri = process.argv[2];
  const resource_io = new ResourceIo(source_uri);
  try {
    const video_reader = Avalanche.createVideoReader();
    await video_reader.init(resource_io);
    const result = await video_reader.extractClipReencode(
      process.argv[3],
      parseFloat(process.argv[4]),
      parseFloat(process.argv[5]),
      (step, total) => {
        log.info('progress', step, total);
      },
    );
    log.info('result', result);
  } catch (err) {
    log.info('failed to extract clip reencode', err);
    return;
  } finally {
    Avalanche.destroy();
  }
  log.info('done');
};

main();
